File: annotation-processor/src/main/java/me/bristermitten/mittenlib/annotations/config/MethodNames.java
package me.bristermitten.mittenlib.annotations.config;

import com.google.inject.Singleton;
import me.bristermitten.mittenlib.annotations.ast.Property;
import me.bristermitten.mittenlib.annotations.util.ElementsFinder;

import javax.inject.Inject;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.TypeElement;
import javax.lang.model.element.VariableElement;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

@Singleton
public class MethodNames {
    private final ElementsFinder elementsFinder;
    private final Map<VariableElement, String> safeNameCache = new HashMap<>();
    private final Map<VariableElement, Set<String>> safeNameCache2 = new HashMap<>();

    @Inject
    MethodNames(ElementsFinder elementsFinder) {
        this.elementsFinder = elementsFinder;
    }

    public String safeMethodName(VariableElement variableElement, TypeElement enclosingClass) {
        return safeNameCache.computeIfAbsent(variableElement,
                elem -> safeMethodName0(elem, enclosingClass));
    }

    public String safeMethodName(Property property) {
        return switch (property.source()) {
            case Property.PropertySource.FieldSource(var field) ->
                    safeMethodName(field, (TypeElement) field.getEnclosingElement());
            case Property.PropertySource.MethodSource(var method) -> method.getSimpleName().toString();
        };
    }

    private String safeMethodName0(VariableElement variableElement, TypeElement enclosingClass) {
        var methodNames = safeNameCache2.computeIfAbsent(variableElement, x -> getNoArgMethodNames(enclosingClass));

        var name = new StringBuilder(variableElement.getSimpleName());
        while (methodNames.contains(name.toString())) {
            name.append("_");
        }
        return name.toString();
    }

    private Set<String> getNoArgMethodNames(TypeElement enclosingClass) {
        var names = new HashSet<String>();
        for (ExecutableElement method : elementsFinder.getAllMethods(enclosingClass)) {
            if (!method.getParameters().isEmpty()) {
                continue;
            }
            names.add(method.getSimpleName().toString());
        }
        return names;
    }
}

File: annotation-processor/src/main/kotlin/me/bristermitten/mittenlib/annotations/config/MethodNames.kt
package me.bristermitten.mittenlib.annotations.config

import com.google.inject.Singleton
import me.bristermitten.mittenlib.annotations.util.ElementsFinder
import javax.inject.Inject
import javax.lang.model.element.TypeElement
import javax.lang.model.element.VariableElement

@Singleton
class MethodNames @Inject internal constructor(private val elementsFinder: ElementsFinder) {
    private val safeNameCache: MutableMap<VariableElement, String> = HashMap()
    private val safeNameCache2: MutableMap<VariableElement, Set<String>> = HashMap()
    fun safeMethodName(variableElement: VariableElement, enclosingClass: TypeElement): String {
        return safeNameCache.computeIfAbsent(
            variableElement
        ) { elem: VariableElement -> safeMethodName0(elem, enclosingClass) }
    }

    private fun safeMethodName0(variableElement: VariableElement, enclosingClass: TypeElement): String {
        val methodNames =
            safeNameCache2.getOrPut(variableElement) { getNoArgMethodNames(enclosingClass) }
        val name = StringBuilder(variableElement.simpleName)
        while (methodNames.contains(name.toString())) {
            name.append("_")
        }
        return name.toString()
    }

    private fun getNoArgMethodNames(enclosingClass: TypeElement): Set<String> {
        val names = HashSet<String>()
        for (method in elementsFinder.getAllMethods(enclosingClass)) {
            if (!method.parameters.isEmpty()) {
                continue
            }
            names.add(method.simpleName.toString())
        }
        return names
    }
}
